### DISCUSSION TITLE
Is there a way to query an input device-specific glyph?

### DISCUSSION THREAD

**Discussion by TheEmbracedOne** [Orange-Panda/Rewired-Glyphs#8] [Q&A] [answered]
I'm looking to display mouse, keyboard and gamepad glyphs. Inputting `<glyph "Cancel">` as per the example on the Github page, it displays mouse right click by default (the glyph is correct) and the GlyphRichTextFormatter changes `<glyph "Cancel">` to `<sprite="mouse-glyphs" name="mouse-glyphs_15">`.

I'd like to specify which input method I'm querying, for example I'd like to do something like `<glyph "Cancel" input="xbox">` or  `<glyph "Cancel" input="keyboard">` to display xbox and keyboard glyphs specifically. Is something like this possible? Thank you

**Comment by Orange-Panda** [comment]
I don't believe this is currently possible but this would be a very useful feature so I will mark this down for a future release

**Comment by Orange-Panda** [comment]
I've implemented this feature on the [development branch](https://github.com/Orange-Panda/Rewired-Glyphs/tree/dev) but I will need to do some additional testing to make sure it is ready for a full release. 

If your use case is really urgent you may be able to try out the experimental feature from there before a stable release.

The syntax works by adding `type=joystick` to the glyph tag. Some examples: `<glyph Jump type=keyboard>` `<glyph Jump type=joystick>` `<glyph Jump type=mouse>`

Be warned that when specifying the controller type it will by design show the `UNBOUND` glyph on unmapped inputs for that controller, even if it exists on another controller.

**Comment by TheEmbracedOne** [comment]
Thank you, I'll get on this asap! I'm working on implementing the controls screen for my game and I'd like to allow rebinding, so showing the right glyph in the right place on the controller markup is very important.

A side question however, would it be possible to specify the controller in type? For example, the default setup has a couple of hardware maps and template maps set up and I'd love to be able to call to a specific map by specifying type, for example for my use case I'm looking to do

<glyph Jump type=mouse>
<glyph Jump type=keyboard>
<glyph Jump type=xbox>
<glyph Jump type=playstation>

in different places. My game supports different gamepads so it'd be great if I could specify which set I'm displaying by, for example, referencing the hardware definition of a hardware map or somesuch. Would this be possible?

**Reply by Orange-Panda** [reply]
By default `InputGlyphs` will show hardware specific symbols for the player's joystick if it recognizes it. 

If you want to enforce a specific symbol type on all hardware you can use the method `InputGlyphs.SetGlyphPreferences`.

The caveat of the above method is that it affects _all_ joystick glyph outputs. If you want to show symbols for the joystick being used by that player `SymbolPreference.Auto` should be sufficient. 

I recognize that being a global preference may be limiting so if your use case is still impossible with those tools I will investigate the possibility of controlling symbol type on a per glyph output basis.

**Comment by TheEmbracedOne** [comment]
Thank you for the info, and this would be good for my use case, as I set joystick icon prefs via a setting (player can set whether they want playstation or xbox icons) regardless of the actual hardware used. I was out today and couldnt get to it yet, will do tomorrow. Is there anything special I'll need to do to get the dev branch update (do I do it via package manager or some other way)?

**Reply by Orange-Panda** [reply, marked as answer]
I've gone ahead and added the ability to specify symbol type on a per rich text glyph basis. You can specify the symbols using `symbol=auto`, `symbol=xbox`, `symbol=ps`, or `symbol=switch`.

For the case you describe where the user can set preferred symbols you may still find the `InputGlyphs.SetGlyphPreferences` method more useful since you won't have to specify for every glyph but you'll have the flexibility to implement whichever way you see fit.

Also no need to worry about switching to the development branch: I have published a new release `1.5.0` which includes these features. Depending on how you installed the package already you can update either:
- (Recommended) Install from the [git url in package manager](https://docs.unity3d.com/Manual/upm-ui-giturl.html) `https://github.com/Orange-Panda/Rewired-Glyphs.git#v1.5.0`
- Or download the source code and import it into your project.

**Reply by TheEmbracedOne** [reply]
Thank you so much, I'll grab this now! Amazing work!

**Reply by TheEmbracedOne** [reply]
I just wanted to say, everything works very nicely. Thank you once again!
